### PR TITLE
Split engine load in to electrical and propulsion

### LINKED ISSUE #12277: Removing the battery wont disable a car's alarm
Well I've never stole a car before, but I'm pretty sure that removing a car's battery should make the alarm stop, if not instantly at least eventually.

Perplexingly, removing the horn does make the alarm stop, even though most car alarms don't rely on the horn exclusively.
## <bountysource-plugin>

Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/12880300-removing-the-battery-wont-disable-a-car-s-alarm?utm_campaign=plugin&utm_content=tracker%2F146201&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F146201&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by mugling**
Starts separating propulsion losses from electrical power generation and intends to fix all current outstanding issues with alternators.

Further preamble to follow

Incidentally fixes #12277

### CODE REVIEW

**Comment by mutability** [comment]
Would it make sense, in the case where batteries + alternators can't meet the power requirement, to shut off energy consumers incrementally until there is enough power available? If I read your current changes right it'll shut off everything, including the engine, if you're even fractionally short on power.

(kinda related: should diesel engines require epower to continue running?)

**Comment by mugling** [comment]
> Would it make sense, in the case where batteries + alternators can't meet the power requirement, to shut off energy consumers incrementally until there is enough power available?

Yes, but I want to focus on a clean implementation here for now. We could/should definitely add your suggestion in a folllow up PR after some consideration about exactly how it would work.

> should diesel engines require epower to continue running

No, they should happily run with out it although for now it does serve reasonably well as simulating engine idle load

**Comment by mugling** [comment]
> Would it make sense, in the case where batteries + alternators can't meet the power requirement, to shut off energy consumers incrementally until there is enough power available?
> 
> Yes, but I want to focus on a clean implementation here for now. We could/should definitely add your suggestion in a folllow up PR after some consideration about exactly how it would work.

Also as engines can both consume and generate (via alternators) electrical power we could end up with accidental strobe-headlights...

**Comment by mutability** [comment]
I was just thinking along the lines of: when you're going through setting enabled = false, update the deficit as you go and break out of the loop when you have enough power. If you don't have enough power after you've tried everything, then battery dies / stall the engine if needed. Emit a message when you turn stuff off, and it stays off until manually reengaged

**Comment by mugling** [comment]
Ready

* Separates engine load into electrical load (via `power_usage`) and propulsion load (`friction_loss`). A further PR will use `power_usage` to calculate the time required to fully charge/discharge the vehicle batteries. A separate PR will rework `friction_loss` and then add unit tests for fuel efficiency

* Alternator load (and therefore fuel consumption) is proportional to electrical demand. Idling engines that have fully recharged their batteries are now considerably more fuel efficient. Next step is to write acceptance tests for fuel consumption at idle for typical vehicles/generators.

* Fixed bug where alternators attached to disable engines would still generate power and drain fuel

* Enabled reactors provide auxiliary electrical power if alternators or local/networked batteries are insufficient. Fully atomic vehicles are now possible and due to the high energy density minireactors are a very suitable powerplant for lightweight electrical vehicles. Each plutonium cell provides 100MJ and is the equivalent to 2.5 storage batteries (40MJ) - this probably needs some later balancing.

File: src/vehicle.cpp [hunk 1]
```diff
@@ -3504,35 +3383,124 @@ int vehicle::charge_battery (int amount, bool include_other_vehicles)
...
+
+    // if insufficient local or network electrical power try to engage any reactors
+    if( amount > 0 && reactor ) {
+        for( auto &pt : parts ) {
+
+            // consider only enabled reactors that have remaining fuel
+            if( pt.is_reactor() && pt.enabled && !pt.is_broken() && pt.ammo_remaining() ) {
+
+                const itype *fuel = item::find_type( pt.ammo_current() );
+                if( fuel->ammo ) {
```
Comment (Coolthulhu) [inline src/vehicle.cpp]: This should be enforced somewhere else - if all reactors need ammo-type ammo, then for a reactor, `ammo_remaining() > 0` should imply `item::find_type( pt.ammo_current() )->ammo != nullptr`.

File: src/vehicle.cpp [hunk 2]
```diff
@@ -3504,35 +3383,124 @@ int vehicle::charge_battery (int amount, bool include_other_vehicles)
...
+        // ...so partial charges have a proportional chance of being added each turn
+        recharge += x_in_y( recharge % 1000, 1000 ) ? 1000 : 0;
+
+        // attempt to recharge battery limiting generated power to total battery capacity
+        generate -= charge_battery( recharge );
+    }
+
+    if( engine_on && eng ) {
+        // calculate load on engine from both propulsion and alternators
+        int load = friction_load() + generate;
```
Comment (Coolthulhu) [inline src/vehicle.cpp]: No efficiency setting?

File: data/json/items/ammo.json [hunk 3]
```diff
@@ -179,6 +179,7 @@
         "weight" : 80,
         "ammo_type" : "plutonium",
         "count" : 5,
+        "energy": 200,
```
Comment (Coolthulhu) [inline data/json/items/ammo.json]: Is that just 4 times as much as gasoline? Is it per volume or unit? If per unit, then gasoline is >50 times more efficient.

Comment (mugling) [inline data/json/items/ammo.json]: It's per effective unit of plutonium where 1 putonium cell = 500 units = 100'000 energy. That plutonium charges multiplier always causes problems but I don't have any better solution at this time

File: src/vehicle.cpp (hunk 2 above)
Comment (mugling) [inline src/vehicle.cpp]: Handled below in the call to `eng.efficiency()`

File: src/vehicle.cpp (hunk 1 above)
Comment (mugling) [inline src/vehicle.cpp]: Ok, I'll downgrade it to an `assert`

File: src/vehicle.cpp (hunk 2 above)
Comment (Coolthulhu) [inline src/vehicle.cpp]: That's engine efficiency, but no generator efficiency. Are they all assumed to be near-100% efficient?

Comment (mugling) [inline src/vehicle.cpp]: Presently yes

**Comment by mugling** [comment]
Incidentally fixes #12277

**Comment by Coolthulhu** [comment]
Check out http://smf.cataclysmdda.com/index.php?topic=13801.0

Apparently electrical devices became very power-hungry because of the power rework.

Is it fixed here? If not, could it be?

**Comment by mugling** [comment]
I've just seen that thread now. I'll take a look

**Comment by mugling** [comment]
Looks like it's already fixed here. After `power_usage()` there is a rescaling by 1000 and then later a multiplication by 6 to convert kW to kJ

**Comment by kevingranade** [comment]
On Nov 23, 2016 2:16 PM, "Oliver Jowett" <notifications@github.com> wrote:

I was just thinking along the lines of: when you're going through setting
enabled = false, update the deficit as you go and break out of the loop
when you have enough power. If you don't have enough power after you've
tried everything, then battery dies / stall the engine if needed. Emit a
message when you turn stuff off, and it stays off until manually reengaged


Outside of a very "smart" car, this isn't going to happen, all current
consumers do so independently, if the system is overloaded everything
browns out or blacks out simultaneously. This is a property of electric
circuits and we shouldn't change it without good reason.

**Comment by mutability** [comment]
It is a playability suggestion, not a realism suggestion. The alternative
would be to give much better feedback that the electrical system was
struggling before it dies entirely.

**Comment by mutability** [comment]
The specific play-ability issue here is that (especially in the early game) running out of battery and stalling the engine is a big deal, restarting it may be impossible. It should not happen without plenty of warning signs.

**Comment by mugling** [comment]
This PR effectively blocks further work on engines - in particular the development of acceptance tests. Can anyone review?

**Comment by Coolthulhu** [comment]
Blazemod complains about REACTOR flag.

**Comment by mugling** [comment]
> Blazemod complains about REACTOR flag.

I'll take a look. What is the complaint?

**Comment by Coolthulhu** [comment]
That it has the flag assigned but base type isn't tool with ammo.

**Comment by Coolthulhu** [comment]
Can't start a fresh car. Complains about lack of gas.

**Comment by mugling** [comment]
Compiling a candidate fix now

**Comment by mugling** [comment]
Yes, that fixed it

**Comment by Coolthulhu** [comment]
What is the expected "translation" ratio between gas and batteries now?
Seems that 1 battery charge costs ~2 gas units (on a Car-brand car).

**Comment by mugling** [comment]
> What is the expected "translation" ratio between gas and batteries now?
> Seems that 1 battery charge costs ~2 gas units (on a Car-brand car).

There isn't a set ratio - it depends on max alternator output and engine epower (ignition requirements)

**Comment by Coolthulhu** [comment]
What would be the perfect ratio? Engine with "always optimal" rpm and no epower consumption.

**Comment by mugling** [comment]
Yes

**Comment by Coolthulhu** [comment]
?

**Comment by Coolthulhu** [comment]
I mean, what would be the ratio on an engine with perfect stats for epower production.

**Comment by mugling** [comment]
I'm not certain

**Comment by Coolthulhu** [comment]
Can't get reactor to work. Changed here or broken earlier?

**Comment by mugling** [comment]
> Can't get reactor to work. Changed here or broken earlier?

Specifically what's wrong?

**Comment by Coolthulhu** [comment]
* Spawn atomic sports car
* Remove storage battery
* Install car battery
* Load the reactor to full
* Doesn't charge
* Restart the reactor
* Still doesn't charge

**Comment by Coolthulhu** [comment]
Oh, looks like it still works despite 0 charge.
Is that how reactors work now?

**Comment by mugling** [comment]
> Is that how reactors work now?

Yes, see preamble

> Enabled reactors provide auxiliary electrical power if alternators or local/networked batteries are insufficient. Fully atomic vehicles are now possible and due to the high energy density minireactors are a very suitable powerplant for lightweight electrical vehicles. Each plutonium cell provides 100MJ and is the equivalent to 2.5 storage batteries (40MJ) - this probably needs some later balancing.

**Comment by Coolthulhu** [comment]
I can't get the reactor to consume plutonium. It grants the power as expected now, but either it has amazing power density or doesn't consume any.

**Comment by mugling** [comment]
It does have fairly epic density. Try dropping the `energy` value for it

**Comment by Coolthulhu** [comment]
OK, got it to actually consume some charges.
The ratio is truly gargantuan - at 2 energy, it consumes ~3 charges (out of 500 per cell) to charge UPS makeshift welder (450 charges) to full. Car battery consumes all 2500 charges to charge said welder to 245. Meaning that one charge of plutonium (not cell) is worth ~1000 battery charges, so one cell is worth 500k battery charges (200 full car batteries).
I guess it's tolerable for it to have overhuge density for now - better too much than too little.

There is one really weird bug I've seen reported on forums - apparently mostly-discharged (to ~450 charges) reactors start draining batteries. I'm not sure if it's testable here, now that reactors are reworked. Will try to unload the reactor and drain it below 400 charges.

**Comment by mugling** [comment]
> EDIT: Fuck, can't unload. Can't use lua (didn't set up codeblocks+lua). Any ideas?

Other than savegame hacking no.

**Comment by Coolthulhu** [comment]
Savegame editing sounds fine.

Found a bug:
Reactor won't power turrets. Once batteries are out, turrets just sit there.

**Comment by mutability** [comment]
> apparently mostly-discharged (to ~450 charges) reactors start draining batteries

This one smells like an overflow.

**Comment by mugling** [comment]
> What would be the perfect ratio? Engine with "always optimal" rpm and no epower consumption.

The latest push prevents an obvious exploit (7.5kW generator with the tiny 2-stroke engine) by capping maximum alternator power to maximum engine output.

The hardcoded effects for diesel and gasoline engines being more damage resistant were dropped for now.

**Comment by Coolthulhu** [comment]
UI problem: when battery runs out, "start engine" is no longer offered, as if it was disabled or broken.

**Comment by mugling** [comment]
> Reactor won't power turrets. Once batteries are out, turrets just sit there.

Should be fixed by 1aabad5

**Comment by mugling** [comment]
> UI problem: when battery runs out, "start engine" is no longer offered, as if it was disabled or broken.

What should happen? Is this critical here as that may involve changes to engine stop/start code which is scary-bad.

**Comment by Coolthulhu** [comment]
> What should happen?

The option should be offered, just so that the player can see it won't work.
Should be offered if the vehicle has one workable engine of any type except muscle, regardless of the state of the engine (except "turned on and working right now", of course).

**Comment by mugling** [comment]
> What should happen?
> 
> The option should be offered, just so that the player can see it won't work.

It's not necessarily trivial to fix that here. It more logically belongs in the PR to replace the engine start code (which is the precursor to multiple engines)

**Comment by mugling** [comment]
Anything further?

**Comment by mugling** [comment]
> UI problem: when battery runs out, "start engine" is no longer offered, as if it was disabled or broken

Latest push (1eac4d5) should fix that

**Comment by Coolthulhu** [comment]
I'm going to test minifridge and solars. If they work, I'm merging.
I'll make an issue documenting overpowered plutonium, but that can wait.

**Comment by Coolthulhu** [comment]
Good enough, I guess.

**Pull request merged**